### DISCUSSION TITLE
Find minimum value in learning queries

### DISCUSSION THREAD

**Discussion by Szpilman2** [UPPAALModelChecker/UPPAAL-Meta#261] [Q&A] [answered]
Dear UPPAAL team,

I trust this message reaches you in good spirits.

I've developed a model comprising two components: software and hardware. (Attached below are their respective images.)
Software Model:
![software](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/dd258ce7-679a-427d-bd58-4b967d4775e5)
Hardware Model:
![hardware](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/a8c094dc-7a63-4799-84cf-d71b9ba27449)

Using the strategy ```minStra = minE(energy) [<=250] : <> software.End``` , I've successfully generated a minimum strategy. However, I'm unsure of how to determine the smallest value of energy within the minStra strategy. Could you please provide guidance on this matter?
In other words, I'm seeking clarification on which query to employ to obtain the minimum value of ```energy``` when my model reaches the state ```software.End``` under minStra strategy? (energy is hybrid clock)

I've also attached my model in here:
[OptEngModel_final.zip](https://github.com/UPPAALModelChecker/UPPAAL-Meta/files/15059531/OptEngModel_final.zip)

Thank you for your assistance.
Best regards,
Peyman

**Comment by mikucionisaau** [comment]
Try value estimation under the strategy:
```
E[<=250; 100] (max:energy) under minStra
```
`max:` records the maximum value of `energy` over individual run. 
Since `energy` is monotonically increasing then `max:energy` will correspond to its last observed value in the run.

Alternatively inspect the energy trajectories:
```
simulate[<=250; 100]{ energy } under minStra
```

I would also suggest to add observation expressions to your learning query. For your current query without observable part of the state UPPAAL will treat as if entire state is observable, which is usually unrealistic and it also drowns the signal with a lot of noise in the learning statistics.

**Reply by Szpilman2** [reply]
thank you very much for answering.
as a matter of fact I would like to get the value of energy when software model reaches to ```End``` state. but the value of ```E[<=250;100] (max: energy) under minStra``` seems to be ```clock bound*energyRate``` (in this example ```250*4 == 1000``` which is not intended for me). in other words I would like to get the min value of energy when software and hardware model are working together (via synchronization). actually I am solving an optimization problem which I have modeled this with hybrid clock. please take a look at my software model:
![Screenshot 2024-04-22 111030](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/bf6d8a72-a230-42fa-8ca8-c8417c4a4345)
I would like to know which one of paths should I take to reach to ```End``` state so that the value of ```energy``` will be minimized.
could you please help me?

also i couldn't find documentation for ```observation expression``` could you possibly help me to know about that?

best regards

**Reply by mikucionisaau** [reply]
Interesting.
You get `clock bound*energyRate` because the `hardware` uses `energy'==4` on ***all*** its locations no matter what, so that's expected. However it will eventually contradict to `energy'==0` in `software.End` and UPPAAL probably does not warn you about it -- that's not good.
I would suggest to replace `energy'==4` with `energy'==power`, where `power` is `int power = 4;` and then update `power` with `power=0` upon moving to `software.End`.

Then I am getting a deterministic value `28` for `energy` in all runs, which is quite boring, but without a strategy there seem to be two outcomes:
![image](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/4868894/1e73923a-592e-4b4a-aa44-8b954e40996b)
so it actually learned something (but not much). 
`A<> software.End` finds a counter-example avoiding `End` and it leads to `software.Error2`, so the strategy avoids the `Error2` somehow. When playing the strategy in concrete simulator, `software` chooses `p1`, so the strategy is not trivial: it does not avoid by going into `p5`, but instead it goes through `p1`, `p2` and `p3` by actively avoiding `Error2`.
![image](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/4868894/9d390c59-b0d4-4206-85c2-af8f09a9dab1)


The observation expressions are in the form of `{ e1, e2 } -> { e3, e4 }` as explained here:
https://docs.uppaal.org/language-reference/query-syntax/learning_queries/

**Reply by mikucionisaau** [reply]
Assuming that `Error` locations need to be avoided by the learning too, you could also set `power` to very high number when `software` reaches those error locations.

**Reply by mikucionisaau** [reply]
To answer your research question: to choose first or second half.
Initially if `hardware` did not yet go into `PowerSaving`, then `software` should choose `p1`.
If `hardware` goes into `PowerSaving` immediately, then `software` can choose either option: `p1` or `p5`.

**Reply by Szpilman2** [reply]
I am deeply appreciative; thank you very much indeed.

In your previous response, you provided the answer to my research question. I am interested in understanding how you arrived at this conclusion, as I aim to replicate your methodology and provide a general rule. Your assistance in this matter would be greatly appreciated.

also I considered all your valuable suggestions about defining a ```power``` variable and the model now looks like this:
![software2](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/009da5d6-d797-44d8-9bc4-0663d65ce114)
![hardware2](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/856d6c10-3731-4700-a903-18a1d36b6f53)


Once again, thank you very much for your help.

**Reply by mikucionisaau** [reply]
In order to examine the strategy choices, you can visit Concrete Simulator tab and choose the strategy at the left bottom corner (the strategy needs to be computed first in Verifier tab to be available in Concrete Simulator).

Then examine the options available in the top left panel: it contains only the transitions allowed by the strategy and other (not to be chosen) transitions are omitted.

![image](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/4868894/a729fb79-628f-4eff-9585-58b8ef488582)

![image](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/4868894/c61f348f-8f71-4565-9c57-28bf41d82505)

![image](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/4868894/ba7dfed3-43a3-40a6-a005-8a521e2a35cb)

The transitions may also have some probabilistic weights associated with them and `Stochastic` can run them based on those weights automatically.

Alternatively, you can save the strategy into a json file using `saveStrategy` query and examine it the choices there.

**Reply by Szpilman2** [reply]
Thank you so much; I truly appreciate your help. Your explanation was excellent and thorough.

I have saved the generated strategy, but I'm having trouble interpreting it. Specifically, I'm unsure how the strategy file determines which half will be chosen. Could you please assist me in understanding this?

also is value of ```28 (result of E[<=250; 100] (max:energy) under minStra)``` correct for minimum energy when we use minimum strategy to get into ```software.End``` location?

Thank you once again for your assistance.

this is my strategy file:
[file.json](https://github.com/UPPAALModelChecker/UPPAAL-Meta/files/15061957/file.json)

Best regards

**Reply by petergjoel** [reply]
The format is encoding a "map of trees". Formally the strategy is represented as a map from actions to  "discrete observables" to functions (represented as trees) over "continuous" observables, finally into "Q-values".

```
S : Act ⇀ ( Disc ⇀ (Cont → ℝ ) )
```

The file is structured as follows (comments inline
```jsonc
{
               "version":1.0,"type":"state->regressor","representation":"map", // meta-data, static at the moment
               "actions":{ // map from index of action used in "regressors" later to text-representation of Uppaal Edge
		"0":"software.p7->software.p7_delay { 1, tau, z := 0 }",
		...,
		"16":"WAIT" // special "NOTHING" action
	},"statevars":[ // observable "Discrete" variable-vector, used as index in "regressors"
		"energy"
	],"pointvars":[ // continious observables. Stratego will regress over these, used in the tree in "regressors"
	],"locationnames":{ ... }, // maping from integer values to text if A.location is observed (A being an instance of a template)
       "regressors": { // actual decision tree. Since the "pointvars" array is empty, this will look like a map
       "(44)": // the statevars-vector observation (44) has the following strategy. I.e. when "energy=44" this is where to find the strategy.
			{"type":"act->point->val","representation":"simpletree","minimize":1, // meta-data defining the format of the regressor. The "minimize"-flag dictates if the minimal or maxmimal value should be chosen as the best action. 
                               "regressor": // the actual decision tree. Since "pointvars" is empty, this is simply a map from actions to Q-values
				{
				"0" : 4, // "software.p7->software.p7_delay { 1, tau, z := 0 }" has the Q-value "4"
				"9" : 0 // "software.p3->software.p4 { y >= 4, use!, y := 0 }" has the Q-value 0 and is thus the best action (according to the strategy) when "energy=44".
				}
			},
....
```

Notice that the Q-values should NEVER be used as having any semantics. They are approximations of the optimal value, but there are quite a few caveats here (among others, discounting which is needed to avoid algebraic loops).

**Reply by Szpilman2** [reply]
Thank you very much for your help. From what I understand, the strategy file provides information about the Q-values. However, I'm uncertain how to determine which of the halves to take based on this file. In other words, how can I use this file to decide whether to proceed from ```software.start``` to ```software.p1``` or ```software.p5```?

```
"(0)":
			{"type":"act->point->val","representation":"simpletree","minimize":1,"regressor":
				{
				"15" : 27.64160000000001,
				"7" : 27.76080077402224
				}
			},
```

I think this part of strategy will help to answer this question. because in the beginning ```energy == 0``` and this part tells us that taking ``` "15":"software.start->software.p1 { 1, tau, 1 }" ``` is the best choice because the Q-value is less than other. am I right?

I would greatly appreciate your assistance in clarifying this matter.

Thank you once again for your support.

**Reply by mikucionisaau** [reply]
In this particular case, the strategy does not really differentiate between "15" and "7", i.e. both actions are almost as good (I don't think there is statistically significant difference). 
If you follow the numbers, then it says pick the smaller one, i.e. "15".

**Reply by Szpilman2** [reply]
Dear Marius,

Hello, I hope this message finds you well,

Previously, I set the ```energy'``` in each location of the hardware TA to be equal to the ```Power``` variable, which is defined as an integer with a value of 4 (```[int Power = 4]```). Now, I want to assign different energy consumption rates for each location in the hardware TA. Specifically, I want the following rates:

- ```state: Active --> energy' == 6```
- ```state: Power Saving --> energy' == 1```
- ```state: Tail --> energy' == 18```
- ```state: Using --> energy' == 14```

However, I'm unsure how to achieve this with the ```Power``` variable. I think I might need an array of power values. You previously mentioned that for learning to occur, the ```Power``` value in the software TA needs to be set to a large number. But with these new requirements (different energy rates in each hardware TA location), I'm not sure how to adjust my model. Could you please guide me on how to implement this? 

Thank you very much.

**Reply by mikucionisaau** [reply, marked as answer]
Simply use the `power` variable in the derivative expression and then the value of the `power` variable can be adjusted on the incoming edges. Uppaal does not really care about the derivative expression as long as it can compute it. Then I presume the `energy` variable is going to be used in the minimization learning query and the automaton location (or `power` variable) to be used in the observable feature part of the query.

The `energy` variable needs to be declared as `hybrid clock` in order to use symbolic (`A[]`, `E<>` etc) queries.
